storage/tests/python/src/norcow.py

```diff
@@ -176,3 +165,166 @@ def _set_sectors(self, data):
 
     def _dump(self):
         return [bytes(sector) for sector in self.sectors]
+
+
+class NorcowBitwise(Norcow):
+    def __init__(self):
+        super().__init__()
+        self.word_size = consts.WORD_SIZE
+        self.magic = consts.NORCOW_MAGIC_AND_VERSION
+        self.item_prefix_len = 4
+        self.lib_name = "libtrezor-storage.so"
+
+    def get_lib_name(self):
+        return self.lib_name
+
+    def is_byte_access(self):
+        return True
+
+    def _is_updatable(self, key: int, new: bytes) -> bool:
+        """
+        Item is updatable if the new value is the same or
+        it changes 1 to 0 only (the flash memory does not
+        allow to flip 0 to 1 unless you wipe it).
+        """
+
+        old, _ = self._find_item(key)
+        if old is None:
+            return False
+        if len(old) != len(new):
+            return False
+        for a, b in zip(old, new):
+            if a & b != b:
+                return False
+        return True
+
+    def _write(self, pos: int, key: int, new_value: bytes) -> int:
+        data = pack("<HH", key, len(new_value)) + align_data(new_value, self.word_size)
+        if pos + len(data) > consts.NORCOW_SECTOR_SIZE:
+            raise RuntimeError("Norcow: item too big")
+        self.sectors[self.active_sector][pos : pos + len(data)] = data
+        return len(data)
+
+    def _norcow_item_length(self, data_len: int) -> int:
+        # APP_ID, KEY_ID, LENGTH, DATA, ALIGNMENT
+        return self.item_prefix_len + data_len + align_int(data_len, self.word_size)
+
+    def _read_item(self, offset: int) -> (int, bytes):
+        if offset >= consts.NORCOW_SECTOR_SIZE:
+            raise ValueError("Norcow: no data on this offset")
+
+        key = self.sectors[self.active_sector][offset : offset + 2]
+        key = int.from_bytes(key, sys.byteorder)
+        if key == consts.NORCOW_KEY_FREE:
+            raise ValueError("Norcow: no data on this offset")
+        length = self.sectors[self.active_sector][offset + 2 : offset + 4]
+        length = int.from_bytes(length, sys.byteorder)
+        value = self.sectors[self.active_sector][offset + 4 : offset + 4 + length]
+
+        return key, value
+
+
+class NorcowBlockwise(Norcow):
+    def __init__(self):
+        super().__init__()
+        self.word_size = 4 * consts.WORD_SIZE
+        self.magic = consts.NORCOW_MAGIC_AND_VERSION + bytes([0x00] * 8)
+        self.item_prefix_len = 4 * consts.WORD_SIZE + 1
+        self.lib_name = "libtrezor-storage-qw.so"
+
+    def get_lib_name(self):
+        return self.lib_name
+
+    def is_byte_access(self):
+        return False
+
+    def _is_updatable(self, key: int, new: bytes) -> bool:
+        """
+        Item is updatable if the new value is the same or
+        it changes 1 to 0 only (the flash memory does not
+        allow to flip 0 to 1 unless you wipe it).
+
+        For flash with no bit access, the item is updatable
```

Comment:
Unfinished comment, probably.

Reply:
and wrong too https://github.com/trezor/trezor-firmware/commit/57a01fb0a49289fd69b1e3fecb1677dc8fcdc91e